File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/py/SocketClient.java
package ch.softappeal.yass.tutorial.py;

import ch.softappeal.yass.remote.*;
import ch.softappeal.yass.serialize.*;
import ch.softappeal.yass.transport.*;
import ch.softappeal.yass.tutorial.contract.*;
import ch.softappeal.yass.tutorial.contract.instrument.*;
import ch.softappeal.yass.tutorial.contract.instrument.stock.*;
import ch.softappeal.yass.tutorial.shared.*;
import ch.softappeal.yass.tutorial.shared.socket.*;
import kotlin.jvm.functions.*;

import java.lang.reflect.*;
import java.util.*;

import static ch.softappeal.yass.transport.MessageSerializerKt.*;
import static ch.softappeal.yass.transport.socket.SocketKt.*;
import static ch.softappeal.yass.transport.socket.SocketTransportKt.*;
import static ch.softappeal.yass.transport.socket.SslSetupKt.*;
import static ch.softappeal.yass.tutorial.contract.Config.*;

public final class SocketClient {

    static Object createObjects() {
        final Node node1 = new Node(1.0);
        final Node node2 = new Node(2.0);
        node1.links.add(node1);
        node1.links.add(node2);
        return Arrays.asList(
            null,
            false,
            true,
            123456,
            -987654,
            1.34545e98d,
            "Hello",
            ">\u0001\u0012\u007F\u0080\u0234\u07FF\u0800\u4321\uFFFF<",
            new byte[]{0, 127, -1, 10, -45},
            new Expiration(2017, 11, 29),
            PriceKind.ASK,
            PriceKind.BID,
            new Stock(123, "YASS", true),
            new UnknownInstrumentsException(Arrays.asList(1, 2, 3)),
            node1
        );
    }

    static final Serializer SERIALIZER = Config.PY_CONTRACT_SERIALIZER;

    static void client(final Client client) {
        final Function3<Method, List<?>, Function0<?>, Object> logger = new Logger(null, Logger.Side.CLIENT);
        final EchoService echoService = client.proxy(PY_ACCEPTOR.echoService);
        final InstrumentService instrumentService = client.proxy(PY_ACCEPTOR.instrumentService, logger);
        System.out.println(echoService.echo("hello"));
        System.out.println(echoService.echo(createObjects()));
        try {
            echoService.echo("exception");
        } catch (final SystemException e) {
            System.out.println(e.details);
        }
        final byte[] big = new byte[1_000_000];
        if (((byte[]) echoService.echo(big)).length != big.length) {
            throw new RuntimeException();
        }
        instrumentService.showOneWay(true, 123);
        System.out.println(instrumentService.getInstruments());
    }

    public static void main(final String... args) {
        client(socketClient(
            new ClientSetup(messageSerializer(SERIALIZER)),
            socketConnector(SocketSetup.ADDRESS, getSocketFactory(SslConfig.CLIENT))
        ));
    }

}

File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/py/SocketClient.kt
package ch.softappeal.yass.tutorial.py

import ch.softappeal.yass.remote.Client
import ch.softappeal.yass.transport.ClientSetup
import ch.softappeal.yass.transport.messageSerializer
import ch.softappeal.yass.transport.socket.socketClient
import ch.softappeal.yass.transport.socket.socketConnector
import ch.softappeal.yass.transport.socket.socketFactory
import ch.softappeal.yass.tutorial.contract.Expiration
import ch.softappeal.yass.tutorial.contract.Node
import ch.softappeal.yass.tutorial.contract.PriceKind
import ch.softappeal.yass.tutorial.contract.PyAcceptor
import ch.softappeal.yass.tutorial.contract.PyContractSerializer
import ch.softappeal.yass.tutorial.contract.SystemException
import ch.softappeal.yass.tutorial.contract.UnknownInstrumentsException
import ch.softappeal.yass.tutorial.contract.instrument.stock.Stock
import ch.softappeal.yass.tutorial.shared.Side
import ch.softappeal.yass.tutorial.shared.SslClient
import ch.softappeal.yass.tutorial.shared.logger
import ch.softappeal.yass.tutorial.shared.socket.Address

fun createObjects(): Any {
    val node1 = Node(1.0)
    val node2 = Node(2.0)
    node1.links.add(node1)
    node1.links.add(node2)
    return listOf(
        null,
        false,
        true,
        123456,
        -987654,
        1.34545e98,
        "Hello",
        ">\u0001\u0012\u007F\u0080\u0234\u07FF\u0800\u4321\uFFFF<",
        byteArrayOf(0, 127, -1, 10, -45),
        Expiration(2017, 11, 29),
        PriceKind.ASK,
        PriceKind.BID,
        Stock(123, "YASS", true),
        UnknownInstrumentsException(listOf(1, 2, 3)),
        node1
    )
}

val Serializer = PyContractSerializer

fun client(client: Client) {
    val echoService = client.proxy(PyAcceptor.echoService)
    val instrumentService = client.proxy(PyAcceptor.instrumentService, logger(null, Side.Client))
    println(echoService.echo("hello"))
    println(echoService.echo(createObjects()))
    try {
        echoService.echo("exception")
    } catch (e: SystemException) {
        println(e.details)
    }
    val big = ByteArray(1_000_000)
    if ((echoService.echo(big) as ByteArray).size != big.size) throw  RuntimeException()
    instrumentService.showOneWay(true, 123)
    println(instrumentService.instruments)
}

fun main(args: Array<String>) {
    client(socketClient(
        ClientSetup(messageSerializer(Serializer)),
        socketConnector(Address, SslClient.socketFactory)
    ))
}
